File: 14_Stack & Queue-II/H460. LFU Cache.java
// LFU Cache
// Least Frequently Used Cache - a very popular DSA + system design interview problem.
// https://leetcode.com/problems/lfu-cache/

// Implementation using Doubly Linked List and HashMap

import java.util.*;

class LFUCache {

    private int capacity;
    private int minFreq;
    private Map<Integer, Integer> keyToVal;
    private Map<Integer, Integer> keyToFreq;
    private Map<Integer, LinkedHashSet<Integer>> freqToKeys;

    public LFUCache(int capacity) {
        this.capacity = capacity;
        this.minFreq = 0;
        this.keyToVal = new HashMap<>();
        this.keyToFreq = new HashMap<>();
        this.freqToKeys = new HashMap<>();
    }

    public int get(int key) {
        if (!keyToVal.containsKey(key)) return -1;
        increaseFreq(key);
        return keyToVal.get(key);
    }

    public void put(int key, int value) {
        if (capacity == 0) return;

        if (keyToVal.containsKey(key)) {
            keyToVal.put(key, value);
            increaseFreq(key);
            return;
        }

        // If full, evict least frequently used
        if (keyToVal.size() == capacity) {
            LinkedHashSet<Integer> keys = freqToKeys.get(minFreq);
            int evictKey = keys.iterator().next(); // oldest inserted among minFreq
            keys.remove(evictKey);
            if (keys.isEmpty()) freqToKeys.remove(minFreq);
            keyToVal.remove(evictKey);
            keyToFreq.remove(evictKey);
        }

        // Insert new key
        keyToVal.put(key, value);
        keyToFreq.put(key, 1);
        freqToKeys.computeIfAbsent(1, k -> new LinkedHashSet<>()).add(key);
        minFreq = 1;
    }

    // Helper: increase frequency of a key
    private void increaseFreq(int key) {
        int freq = keyToFreq.get(key);
        keyToFreq.put(key, freq + 1);

        // Remove from old freq list
        LinkedHashSet<Integer> keys = freqToKeys.get(freq);
        keys.remove(key);
        if (keys.isEmpty()) {
            freqToKeys.remove(freq);
            if (freq == minFreq) minFreq++;
        }

        // Add to new freq list
        freqToKeys
            .computeIfAbsent(freq + 1, k -> new LinkedHashSet<>())
            .add(key);
    }

    public static void main(String[] args) {
        LFUCache cache = new LFUCache(2);
        cache.put(1, 1);
        cache.put(2, 2);
        System.out.println(cache.get(1)); // returns 1
        cache.put(3, 3); // evicts key 2
        System.out.println(cache.get(2)); // returns -1 (not found)
        System.out.println(cache.get(3)); // returns 3
        cache.put(4, 4); // evicts key 1
        System.out.println(cache.get(1)); // returns -1 (not found)
        System.out.println(cache.get(3)); // returns 3
        System.out.println(cache.get(4)); // returns 4
    }
}

File: JavaScript/14_Stack & Queue-II/H460. LFU Cache.js
// 460. LFU Cache
// https://leetcode.com/problems/lfu-cache/

/**
 * Creates an LFU Cache (functional approach)
 * @param {number} capacity - Maximum capacity
 * @returns {Object} LFUCache operations
 */
function createLFUCache(capacity) {
    const keyToVal = new Map();
    const keyToFreq = new Map();
    const freqToKeys = new Map();
    let minFreq = 0;

    function updateFreq(key) {
        const freq = keyToFreq.get(key);
        keyToFreq.set(key, freq + 1);
        
        // Remove from current frequency list
        freqToKeys.get(freq).delete(key);
        if (freqToKeys.get(freq).size === 0) {
            freqToKeys.delete(freq);
            if (minFreq === freq) minFreq++;
        }
        
        // Add to new frequency list
        if (!freqToKeys.has(freq + 1)) {
            freqToKeys.set(freq + 1, new Set());
        }
        freqToKeys.get(freq + 1).add(key);
    }

    function get(key) {
        if (!keyToVal.has(key)) return -1;
        updateFreq(key);
        return keyToVal.get(key);
    }

    function put(key, value) {
        if (capacity === 0) return;
        
        if (keyToVal.has(key)) {
            keyToVal.set(key, value);
            updateFreq(key);
            return;
        }
        
        if (keyToVal.size >= capacity) {
            // Remove LFU key (first in minFreq set)
            const keysAtMinFreq = freqToKeys.get(minFreq);
            const keyToRemove = keysAtMinFreq.values().next().value;
            keysAtMinFreq.delete(keyToRemove);
            if (keysAtMinFreq.size === 0) {
                freqToKeys.delete(minFreq);
            }
            keyToVal.delete(keyToRemove);
            keyToFreq.delete(keyToRemove);
        }
        
        keyToVal.set(key, value);
        keyToFreq.set(key, 1);
        if (!freqToKeys.has(1)) {
            freqToKeys.set(1, new Set());
        }
        freqToKeys.get(1).add(key);
        minFreq = 1;
    }

    return { get, put };
}

// Test cases
const lfu = createLFUCache(2);
lfu.put(1, 1);
lfu.put(2, 2);
console.log(lfu.get(1));      // 1
lfu.put(3, 3);                // evicts key 2
console.log(lfu.get(2));      // -1
console.log(lfu.get(3));      // 3
lfu.put(4, 4);                // evicts key 1
console.log(lfu.get(1));      // -1
console.log(lfu.get(3));      // 3
console.log(lfu.get(4));      // 4
